hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManagerStarter.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.tracing.TracingUtil;
+import org.apache.hadoop.ozone.common.StorageInfo;
+import org.apache.hadoop.util.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+
+import java.io.IOException;
+
+import static org.apache.hadoop.util.ExitUtil.terminate;
+
+/**
+ * This class provides a command line interface to start the SCM
+ * using Picocli.
+ */
+
+@Command(name = "ozone scm",
+    hidden = true, description = "Start or initialize the scm server.",
+    versionProvider = HddsVersionProvider.class,
+    mixinStandardHelpOptions = true)
+public class StorageContainerManagerStarter extends GenericCli {
+
+  private OzoneConfiguration conf;
+  private SCMStarterInterface receiver;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(StorageContainerManagerStarter.class);
+
+  public static void main(String[] args) throws Exception {
+    TracingUtil.initTracing("StorageContainerManager");
+    new StorageContainerManagerStarter(
+        new StorageContainerManagerStarter.SCMStarterHelper()).run(args);
+  }
+
+  public StorageContainerManagerStarter(SCMStarterInterface receiverObj) {
+    super();
+    receiver = receiverObj;
+  }
+
+  @Override
+  public Void call() throws Exception {
+    commonInit();
+    startScm();
+    return null;
+  }
+
+  /**
+   * This function implements a sub-command to generate a new
+   * cluster ID from the command line.
+   */
+  @CommandLine.Command(name = "--genclusterid",
+      customSynopsis = "ozone scm [global options] --genclusterid [options]",
+      hidden = false,
+      description = "Generate a new Cluster ID",
+      mixinStandardHelpOptions = true,
+      versionProvider = HddsVersionProvider.class)
+  public void generateClusterId() {
+    commonInit();
+    System.out.println("Generating new cluster id:");
+    System.out.println(receiver.generateClusterId());
+  }
+
+  /**
+   * This function implements a sub-command to allow the SCM to be
+   * initialized from the command line.
+   *
+   * @param clusterId - Cluster ID to use when initializing. If null,
+   *                  a random ID will be generated and used.
+   */
+  @CommandLine.Command(name = "--init",
+      customSynopsis = "ozone scm [global options] --init [options]",
+      hidden = false,
+      description = "Initialize / format the SCM",
+      mixinStandardHelpOptions = true,
+      versionProvider = HddsVersionProvider.class)
+  public void initScm(@CommandLine.Option(names = { "--clusterid" },
+      description = "Optional: The cluster id to use when formatting SCM",
+      paramLabel = "id") String clusterId)
+      throws Exception {
+    commonInit();
+    boolean result = receiver.init(conf, clusterId);
+    if (!result) {
+      throw new IOException("SCM Init failed. Check the log for more details");
```

Comment:
print the exception below ? Or is it just a bad experience ? @elek ? @arp7 ?

Reply:
+1 for printing the exception.

Reply:
The method StorageContainerManager.init() which actually does the init(), looks like this:

```
  /**
   * Routine to set up the Version info for StorageContainerManager.
   *
   * @param conf OzoneConfiguration
   * @return true if SCM initialization is successful, false otherwise.
   * @throws IOException if init fails due to I/O error
   */
  public static boolean scmInit(OzoneConfiguration conf,
      String clusterId) throws IOException {
    SCMStorageConfig scmStorageConfig = new SCMStorageConfig(conf);
    StorageState state = scmStorageConfig.getState();
    if (state != StorageState.INITIALIZED) {
      try {
        if (clusterId != null && !clusterId.isEmpty()) {
          scmStorageConfig.setClusterId(clusterId);
        }
        scmStorageConfig.initialize();
        ...
        return true;
      } catch (IOException ioe) {
        LOG.error("Could not initialize SCM version file", ioe);
        return false;
      }
    } else {
      System.out.println(
          "SCM already initialized. Reusing existing"
              + " cluster id for sd="
              + scmStorageConfig.getStorageDir()
              + ";cid="
              + scmStorageConfig.getClusterID());
      return true;
    }
  }
```
Note the Java doc states it throws IOException, but it doesn't really (at least in some circumstances) as IOE is caught and then false is returned rather than propagating the exception to the caller. I think it would be better if this init() method throws the exception, rather than returning false and this would let us print the exception on the CLI rather than this message. What do you think?

Reply:
yes, makes sense. If the init failed it is good to see the error in its full glory.

Reply:
OK - I will change the storageContainerManager.scmInit() method so it logs and then re-throws the exception - that will let me simplify the starter code too.

Reply:
Looking at this a bit more, is it valid to assume that when "ozone scm --init" is called, the log will go to the console, rather than hidden in a log file?

The way we have Pico CLI configured, is that if you throw an exception in a command, it causes the process to have a non-zero exit status. If the exception has a cause, it prints just the cause to stder and if it has no cause, it prints the entire stack trace to stderr.

So the output from the current implementation would be:

```
...
2019-06-06 18:13:52,696 INFO  server.StorageContainerManagerStarter (LogAdapter.java:info(51)) - registered UNIX signal handlers for [TERM, HUP, INT]
2019-06-06 18:13:52,827 WARN  server.ServerUtils (ServerUtils.java:getScmDbDir(145)) - ozone.scm.db.dirs is not configured. We recommend adding this setting. Falling back to ozone.metadata.dirs instead.
SCM initialization succeeded.Current cluster id for sd=/tmp/metadata/scm;cid=CID-6786a617-9e83-4135-996f-749c3e07dd8a
2019-06-06 18:13:52,868 ERROR server.StorageContainerManager (StorageContainerManager.java:scmInit(608)) - Could not initialize SCM version file
java.io.IOException
	at org.apache.hadoop.hdds.scm.server.StorageContainerManager.scmInit(StorageContainerManager.java:606)
	at org.apache.hadoop.hdds.scm.server.StorageContainerManagerStarter$SCMStarterHelper.init(StorageContainerManagerStarter.java:146)
	at org.apache.hadoop.hdds.scm.server.StorageContainerManagerStarter.initScm(StorageContainerManagerStarter.java:105)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at picocli.CommandLine.execute(CommandLine.java:1189)
	at picocli.CommandLine.access$800(CommandLine.java:141)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:1367)
	<snipped>
org.apache.hadoop.hdds.scm.server.StorageContainerManagerStarter.main(StorageContainerManagerStarter.java:56)
SCM Init failed. Check the log for more details
2019-06-06 18:13:52,875 INFO  server.StorageContainerManagerStarter (LogAdapter.java:info(51)) - SHUTDOWN_MSG: 
/************************************************************
SHUTDOWN_MSG: Shutting down StorageContainerManager at SOdonnell-MBP15.local/192.168.0.20
************************************************************/
```

So perhaps I should just change the IOE cause to "SCM Init failed." and leave the rest as is based on the above output? Otherwise, we will get the same stack printed to the console twice.

Reply:
yes,  makes sense. Let us not repeat the same message twice on the console.